File: metodo_da_bisseccao.py
'''
    Exemplo usado: x^3-x-1 = 0
'''
import matplotlib.pyplot as plt
import math
import numpy

def metodo_da_bisseccao ():

    n = 0
    nmax = 100
    tol = 1*10**(-12)

    intervalo = [1.0, 1.5]
    x = []
    a = []
    b = []

    a.append (intervalo[0]), b.append (intervalo[1])

    print (f'Tolerância: {tol:.12f}')
    print (f'Intervalo usado: [{a[0]}, {b[0]}]')
    
    if (func(a[0]) * func(b[0]) > 0):
        print (f'Não temos raiz no intervalo: [{a[0]}, {b[0]}], por favor escolha outro!')

    while True:
        x.append(round(0.5 * (a[n] + b[n]), 16))

        if (abs (func(x[n])) < tol):
            print (f'O programa convergiu na interação {n} com x = {x[n]:.16f}')
            break
        else:
            if (func (a[n]) * func (x[n]) < 0):
                a.append(round(a[n], 16))
                b.append(round(x[n], 16))
            else:
                b.append(round (b[n], 16))
                a.append(round (x[n], 16))
        n += 1
        if (n == nmax):
            print ('O programa terminou sem atingit a tolerância desejada')
            break
    
    for i in range (0, n):
        if (i < 10):
            print (f' Interação: {i: 5} | x[0{i}] = {x[i]: <25.16f} | x[0{i}] = {a[i]: <20.16f} | b[0{i}] = {b[i]: <20.16f} | Erro: {abs(a[i]-b[i])/2:.16f}')
        else:
            print (f' Interação: {i: 5} | x[{i}] = {x[i]: <25.16f} | x[{i}] = {a[i]: <20.16f} | b[{i}] = {b[i]: <20.16f} | Erro: {abs(a[i]-b[i])/2:.16f}')

    # Plotagem do gráfico
    i = [i for i in range (0, len(x))]
    plt.title ("Método da Bissecção")
    plt.xlabel ("Eixo X")
    plt.ylabel ("Eixto Y")
    plt.plot (i, x)
    plt.show ()

def func (x):
    return x**3 - x + 1

if __name__ == "__main__":
    metodo_da_bisseccao ()

File: metodo_da_bisseccao.f90
! Exemplo usado: x^3-x-1 = 0
program bissecao
    implicit none

    ! Declaração de variáveis
    integer :: k, sair, nmax, i
    double precision, dimension (:), allocatable :: x, a, b
    double precision :: tol

    nmax = 100

    allocate(x(0:nmax))
    allocate(a(0:nmax))
    allocate(b(0:nmax))

    tol = 1.d-12
    sair = 0

    k = 0
    a(0) = 1.0d0
    b(0) = 1.5d0

    ! Verificação de f(a) * f(b) > 0
    if (f(a(0))*f(b(0)).gt.0) then
        write (*,*) 'Não temos raiz nesse intervalo, escolha outro'
        sair = 1
    end if

    do while (sair.eq.0)
        x(k) = 0.5d0*(a(k)+b(k))

        if (abs(f(x(k))).lt.tol) then 
            write (*,*) 'A raiz foi encontrada e o programa convergiu'
            sair = 1
        else
            if (f(a(k))*f(x(k)).lt.0d0) then 
                a(k+1) = a(k)
                b(k+1) = x(k)
            else 
                b(k+1) = b(k)
                a(k+1) = x(k)
            end if
        end if
        k = k + 1

        if (k.eq.nmax) then
            write (*,*) 'Foi atingido o número máximo de interações e o programa não convergiu'
            sair = 1
        end if
    end do

    open (unit = 123, file = 'saida_bis.dat', status = 'unknown')
    
    do i = 0, k-1
        write (123,*) i, x(i), a(i), b(i), abs(a(i) - b(i))/2.d0
    end do
    close (unit = 123)

    deallocate(x)
    deallocate(b)
    deallocate(a)
    

    contains

        function f(x)
            
            implicit none

            double precision :: f, x

            f = x**3-x-1.d0

            return

        end function f

end program bissecao